rust/src/core/common/converter.rs

```diff
@@ -0,0 +1,1009 @@
+
+/// A type that can be constructed from (part of) an ArbData object.
+///
+/// Any data not consumed by the construction stays in the ArbData object for
+/// further interpretation.
+pub trait FromArb
+where
+    Self: std::marker::Sized,
+{
+    /// Construct from the given ArbData, taking the parameters used for
+    /// construction out of the data object.
+    fn from_arb(src: &mut ArbData) -> Result<Self>;
+}
+
+impl FromArb for () {
+    fn from_arb(_: &mut ArbData) -> Result<Self> {
+        Ok(())
+    }
+}
+
+impl FromArb for u64 {
+    fn from_arb(src: &mut ArbData) -> Result<Self> {
+        let mut args = src.get_args_mut().drain(..1);
+        let i = args
+            .nth(0)
+            .ok_or_else(oe_inv_arg("expected 64-bit integer argument in ArbData"))?;
+        Ok(u64::from_le_bytes(i[..].try_into().ok().ok_or_else(
+            oe_inv_arg("expected 64-bit integer argument in ArbData"),
+        )?))
+    }
+}
+
+impl FromArb for f64 {
+    fn from_arb(src: &mut ArbData) -> Result<Self> {
+        let mut args = src.get_args_mut().drain(..1);
+        let f = args
+            .nth(0)
+            .ok_or_else(oe_inv_arg("expected double argument in ArbData"))?;
+        Ok(f64::from_le_bytes(f[..].try_into().ok().ok_or_else(
+            oe_inv_arg("expected double argument in ArbData"),
+        )?))
+    }
+}
+
+impl FromArb for (f64, f64, f64) {
+    fn from_arb(src: &mut ArbData) -> Result<Self> {
+        let a = f64::from_arb(src)?;
+        let b = f64::from_arb(src)?;
+        let c = f64::from_arb(src)?;
+        Ok((a, b, c))
+    }
+}
+
+impl FromArb for Matrix {
+    fn from_arb(src: &mut ArbData) -> Result<Self> {
+        let mut args = src.get_args_mut().drain(..1);
+        let data = args
+            .nth(0)
+            .ok_or_else(oe_inv_arg("expected matrix argument in ArbData"))?;
+        if data.len() % 16 != 0 {
+            inv_arg("invalid matrix size")?;
+        }
+        let num_entries = data.len() / 16;
+        if num_entries != num_entries.integer_sqrt().pow(2) {
+            inv_arg("invalid matrix size")?;
+        }
+        let mut entries = Vec::with_capacity(num_entries);
+        for i in 0..num_entries {
+            let re = f64::from_le_bytes(data[i * 16..i * 16 + 8].try_into().unwrap());
+            let im = f64::from_le_bytes(data[i * 16 + 8..i * 16 + 16].try_into().unwrap());
+            entries.push(Complex64::new(re, im));
+        }
+        Ok(entries.into())
+    }
+}
+
+/// A type that can be converted into (part of) an ArbData object.
+///
+/// This is the reverse of `FromArb`. Generally, constructing an arb from a
+/// number of `ToArb` objects in a certain order should make the objects
+/// recoverable using `FromArb` in the reverse order, assuming all involved
+/// types implement both directions.
+pub trait ToArb {
+    /// Convert self to ArbData parameters and add them to the data object.
+    fn to_arb(self, dest: &mut ArbData);
```

Comment:
Is there a reason why you don't return the `&mut ArbData`?

Reply:
Also, does it have to take `self`?

Reply:
> Is there a reason why you don't return the `&mut ArbData`?

Not sure why one would do that?

> Also, does it have to take `self`?

Not to satisfy the borrow checker, reasoning is that semantically the value is moved into the `ArbData`.